ocaml/xapi/repository.ml

```diff
@@ -0,0 +1,116 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module Unixext = Xapi_stdext_unix.Unixext
+
+module D = Debug.Make (struct let name = "repository" end)
+
+open D
+
+let reposync_mutex = Mutex.create ()
+
+let create_repository_record ~__context ~name_label ~name_description ~binary_url ~source_url =
+  let ref = Ref.make () in
+  let uuid = Uuidm.to_string (Uuidm.create `V4) in
+  Db.Repository.create
+    ~__context
+    ~ref
+    ~uuid
+    ~name_label
+    ~name_description
+    ~binary_url
+    ~source_url
+    ~hash:""
+    ~up_to_date:false;
+  ref
+
+let assert_url_is_valid ~url =
+  try
+    let uri = Uri.of_string url in
+    match Uri.scheme uri with
+    | Some "http" | Some "https" -> 
+      begin match Uri.host uri with
+        | Some h ->
+          begin match !Xapi_globs.repository_domain_name_allowlist with
+            | [] -> ()
+            | l ->
+              begin match List.exists (fun d -> Astring.String.is_suffix ("." ^ d) h) l with
+                | true -> ()
+                | false ->
+                  let msg = "host is not in allowlist" in
+                  raise Api_errors.(Server_error (internal_error, [msg]))
+              end
+          end
+        | None -> raise Api_errors.(Server_error (internal_error, ["invalid host in url"]))
+      end
+    | _ -> raise Api_errors.(Server_error (internal_error, ["invalid scheme in url"]))
```

Comment:
Can we put the url in these error messages to make debugging easier? Alternatively we could just log it

Reply:
These internal errors will be caught below and raise a invalid_base_url which is exposed to users. In this API error, the url will be there as a parameter.